pkg/runner/cluster_k8s.go

```diff
@@ -293,6 +295,131 @@ func (*ClusterK8sRunner) ID() string {
 	return "cluster:k8s"
 }
 
+func (c *ClusterK8sRunner) healthcheckK8s() (k8sCheck api.HealthcheckItem) {
+	k8sCheck = api.HealthcheckItem{Name: "k8s", Status: api.HealthcheckStatusOK, Message: "k8s cluster is running"}
+	err := exec.Command("sh", "-c", "kops validate cluster").Run()
+	if err != nil {
+		k8sCheck = api.HealthcheckItem{Name: "k8s", Status: api.HealthcheckStatusFailed, Message: fmt.Sprintf("k8s cluster validation failed: %s", err)}
+		return
+	}
+	return
+}
+
+func (c *ClusterK8sRunner) healthcheckEFS() (efsCheck api.HealthcheckItem) {
+	efsCheck = api.HealthcheckItem{Name: "efs", Status: api.HealthcheckStatusFailed, Message: "efs provisioner is not running"}
+
+	client := c.pool.Acquire()
+	defer c.pool.Release(client)
+
+	res, err := client.CoreV1().Pods(c.config.Namespace).List(metav1.ListOptions{
+		LabelSelector: "app=efs-provisioner",
+	})
+	if err != nil {
+		return
+	}
+	if len(res.Items) != 1 {
+		return
+	}
+
+	pod := res.Items[0]
+	if pod.Status.Phase != "Running" {
+		return
+	}
+
+	efsCheck = api.HealthcheckItem{Name: "efs", Status: api.HealthcheckStatusOK, Message: "efs provisioner is running"}
+	return
+}
+
+func (c *ClusterK8sRunner) healthcheckRedis() (redisCheck api.HealthcheckItem) {
+	redisCheck = api.HealthcheckItem{Name: "redis", Status: api.HealthcheckStatusFailed, Message: "redis service is not running"}
+
+	client := c.pool.Acquire()
+	defer c.pool.Release(client)
+
+	res, err := client.CoreV1().Pods(c.config.Namespace).List(metav1.ListOptions{
+		LabelSelector: "app=redis",
+	})
+	if err != nil {
+		return
+	}
+	if len(res.Items) != 1 {
+		return
+	}
+
+	pod := res.Items[0]
+	if pod.Status.Phase != "Running" {
+		return
+	}
+
+	redisCheck = api.HealthcheckItem{Name: "redis", Status: api.HealthcheckStatusOK, Message: "redis service is running"}
+	return
+}
+
+func (c *ClusterK8sRunner) healthcheckSidecar() (sidecarCheck api.HealthcheckItem) {
+	sidecarCheck = api.HealthcheckItem{Name: "sidecar", Status: api.HealthcheckStatusFailed, Message: "sidecar service is not reachable"}
+
+	// get number of worker nodes
+	client := c.pool.Acquire()
+	defer c.pool.Release(client)
+
+	res, err := client.CoreV1().Nodes().List(metav1.ListOptions{
+		LabelSelector: "kubernetes.io/role=node",
+	})
+	if err != nil {
+		return
+	}
+
+	nodes := len(res.Items)
+
+	pods, err := client.CoreV1().Pods(c.config.Namespace).List(metav1.ListOptions{
+		LabelSelector: "name=testground-sidecar",
+	})
+	if err != nil {
+		return
+	}
+	if len(pods.Items) != nodes {
+		return
+	}
+
+	for _, pod := range pods.Items {
+		if pod.Status.Phase != "Running" {
+			return
+		}
+	}
+
+	sidecarCheck = api.HealthcheckItem{Name: "sidecar", Status: api.HealthcheckStatusOK, Message: "sidecar service is running"}
+	return
+}
+
+func (c *ClusterK8sRunner) Healthcheck(fix bool, engine api.Engine, writer io.Writer) (*api.HealthcheckReport, error) {
+
+	// TODO: initialize the runner correctly.
```

Comment:
@raulk @Robmat05 

This is obviously not the right thing to do.

DoRun in engine.go runs the healthcheck for the build and run. This is poorly implemented here. The "fix" is unused, and because the runner is uninitialized, even the checks don't work the first time it's tried.

If this is to go in 0.3, I will have to focus on this to tackle these two points. I would also be okay with opening two tickets as separate items
1. Implement healthcheck fix on k8s
2. come up  with a better k8s runner init.

Reply:
Yeah, the init logic needs to be untangled. It ends up initialising a singleton k8s runner with the configuration it first sees. But that config is really a _run-specific_ configuration, so it's incorrect to globally initialise the runner with it.

IMO, what should happen is:

* Healthcheck checks and/or initialises a pool (fix action) by picking up the Kubernetes config from the env (`Engine.EnvConfig()`), or defaulting to the default configuration.
* The `podResourceCPU` and `podResourceMemory` state variables that are populated on init from the first run configuration we see should not be in the state of the runner. They should be calculated per-run.
* I don't understand why `c.maxAllowedPods` is even in the state of the runner -- it's calculated every time a run happens, and it's racy under a concurrent scenario.